### PR TITLE
Add scala syntax highlights

### PR DESCRIPTION
**Pull request by dwat3r**
Hi!

I really like helix, so I'd like to add scala syntax highlight and LSP support for it.
Amazingly, LSP support with metals works out-of-the-box, although additional metals commands like import build and others obviously needs to be added. 

I've copied the tree-sitter highlights and injections from nvim-treesitter, but found a couple of issues with it, and couldn't figure out it yet, so the syntax highlighting is partial currently. I'm happy to receive suggestions how to migrate the whole from nvim.

### CODE REVIEW

File: runtime/queries/scala/highlights.scm [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+
+;; special keywords
+
+"new" @keyword.operator
+
+[
+  "else"
+  "if"
+  "match"
+] @conditional
```
Comment (archseer) [inline runtime/queries/scala/highlights.scm:160]: @keyword.conditional

File: runtime/queries/scala/highlights.scm [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+; ] @keyword.control.exception
+
+; "return" @keyword.control.return
+
+(comment) @comment
+
+;; `case` is a conditional keyword in case_block
+
+(case_block
+  (case_clause ("case") @conditional))
```
Comment (archseer) [inline runtime/queries/scala/highlights.scm:206]: `@keyword.conditional`

**Review by archseer — CHANGES_REQUESTED** [review]

**Comment by amousset** [comment]
I checked open PR before starting https://github.com/helix-editor/helix/pull/1279 and it happened to be the day another identical PR is opened :smile: 

To make syntax highlighting work I had to update the tree-sitter-scala submodule.

File: runtime/queries/scala/highlights.scm [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+
+; method invocation
+
+
+(call_expression
+  function: (identifier) @function)
+
+(call_expression
+  function: (field_expression
+    field: (identifier) @method))
```
Comment (archseer) [inline runtime/queries/scala/highlights.scm:72]: `@function.method`

File: runtime/queries/scala/highlights.scm [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+  (#match? @function.builtin "^super$")
+)
+
+; function definitions
+
+(function_definition
+  name: (identifier) @function)
+
+(parameter
+  name: (identifier) @parameter)
```
Comment (archseer) [inline runtime/queries/scala/highlights.scm:92]: `@variable.parameter`

File: runtime/queries/scala/highlights.scm [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+
+; expressions
+
+
+(field_expression field: (identifier) @property)
```
Comment (archseer) [inline runtime/queries/scala/highlights.scm:97]: `@variable.other.member`

File: runtime/queries/scala/highlights.scm [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+(infix_type operator: (operator_identifier) @operator)
+(infix_type operator: (operator_identifier) @operator)
+
+; literals
+(boolean_literal) @constant.builtin.boolean
+(integer_literal) @constant.numeric.integer
+(floating_point_literal) @constant.numeric.float
+
+[
+(symbol_literal)
```
Comment (archseer) [inline runtime/queries/scala/highlights.scm:112]: `symbol_literal` could be `@string.special.symbol`

File: runtime/queries/scala/highlights.scm [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+] @keyword.control.conditional
+
+[
+ "("
+ ")"
+ "["
+ "]"
+ "{"
+ "}"
+]  @punctuation.bracket
```
Comment (archseer) [inline runtime/queries/scala/highlights.scm:169]:
```suggestion
] @punctuation.bracket
```

File: runtime/queries/scala/highlights.scm [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+
+"def" @keyword.function
+
+[
+  "=>"
+  "<-"
+  "@"
+] @keyword.operator
+
+"import" @include
```
Comment (archseer) [inline runtime/queries/scala/highlights.scm:191]: `@keyword.control.import`

File: runtime/queries/scala/highlights.scm [hunk 9]
```diff
@@ -0,0 +1,200 @@
...
+
+"return" @keyword.control.return
+
+(comment) @comment
+
+;; `case` is a conditional keyword in case_block
+
+(case_block
+  (case_clause ("case") @keyword.conditional))
```
Comment (archseer) [inline runtime/queries/scala/highlights.scm:200]:
```suggestion
  (case_clause ("case") @keyword.control.conditional))
```

File: runtime/queries/scala/highlights.scm [hunk 10]
```diff
@@ -0,0 +1,200 @@
...
+(wildcard) @keyword
+
+;; special keywords
+
+"new" @keyword.operator
+
+[
+  "else"
+  "if"
+  "match"
```
Comment (archseer) [inline runtime/queries/scala/highlights.scm:156]:
```suggestion
 "else"
 "if"
 "match"
```

**Review by archseer — CHANGES_REQUESTED** [review]

**Comment by dwat3r** [comment]
> I checked open PR before starting #1279 and it happened to be the day another identical PR is opened smile
> 
> To make syntax highlighting work I had to update the tree-sitter-scala submodule.
Thanks, this was also required to make scala syntax highlight work :) 

I've corrected the syntax, but I'm not sure why helix doesn't color things like objects and other stuff like neovim with the nvim-treesitter.

Here's a comparison screenshot (both are using their own onedark theme (hx: builtin, nvim: ful1e5/onedark.nvim)):
With neovim treesitter:
scala-nvim
helix with this PR:
scala-helix

Also apologies for spoiling some old advent of code for you :D

**Comment by archseer** [comment]
It's probably the theme definitions not being defined for every syntax scope. I actually prefer it this way because it strikes a nice balance (is it really necessary to highlight **_every_** word? Not a single word in the first image isn't highlighted..)

I'll do some debugging locally though.

File: runtime/queries/scala/highlights.scm [hunk 11]
```diff
@@ -0,0 +1,202 @@
+; CREDITS @stumash (stuart.mashaal@gmail.com)
+
+;; variables
+
+(identifier) @variable
```
Comment (archseer) [inline runtime/queries/scala/highlights.scm:5]: Move this to the end of the file. tree-sitter takes the first matching query, and this matches and overrides a lot of the later queries.

It significantly improved highlighting for me: 
pos

File: runtime/queries/scala/highlights.scm [hunk 12]
```diff
@@ -0,0 +1,202 @@
...
+(infix_type operator: (operator_identifier) @operator)
+(infix_type operator: (operator_identifier) @operator)
+
+; literals
+(boolean_literal) @constant.builtin.boolean
+(integer_literal) @constant.numeric.integer
+(floating_point_literal) @constant.numeric.float
+
+
+(symbol_literal) @string_special_symbol
```
Comment (archseer) [inline runtime/queries/scala/highlights.scm:112]: `@string.special.symbol`

**Comment by dwat3r** [comment]
You're right, too much highlight is not very good. Although I've missed the functions being highlighted, but your fix made it perfect! :)

**Pull request merged**